active-rfcs/0000-router-view-keep-alive-transitions.md

```diff
@@ -0,0 +1,101 @@
+- Start Date: 2020-04-20
+- Target Major Version: Vue Router 4.x
+- Reference Issues:
+- Implementation PR:
+
+# Summary
+
+Given the changes to functional components in Vue 3, `KeepAlive` and `Transition` usage combined with `RouterView` is no longer possible by simply wrapping `RouterView` with `KeepAlive`/`Transition`. Instead we need a way to directly provide the component rendered by `RouterView` to those components:
+
+```vue
+<router-view v-slot="{ Component, attrs }">
+  <transition :name="transitionName" mode="out-in">
+    <component v-if="Component" :is="Component" v-bind="attrs"></component>
+  </transition>
+</router-view>
+```
+
+# Motivation
+
+As described at https://github.com/vuejs/vue-next/issues/906#issuecomment-611080663, we need a new API to allow the usage of `KeepAlive` and other components accepting slots with `RouterView`. In order to implement such behavior we need to be able to **directly** wrap the component rendered by `RouterView`. The only way to do so by having access to the component rendered by `RouterView` and the _props_ passed to it.
+
+# Detailed design
+
+We can achieve this level of control through a slot:
+
+```vue
+<router-view v-slot="{ Component, attrs }">
+  <component :is="Component" v-bind="attrs"></component>
+</router-view>
+```
+
+In the example, `Component` is the _component definition_ that can be passed to the function `h` or to the `is` prop of `component`. `attrs` is a set of event listeners and props that `RouterView` **requires** in order to work properly. Namely, `onVnodeMounted`/`onVnodeUnmounted` and a `ref`. It also contains any _attribute_/_prop_ passed to `RouterView`. Let's say we pass a `class` to it, we will find it in `attrs`:
+
+```vue
+<router-view class="view" v-slot="{ attrs }">
+  {{ attrs.class }} -> view
+</router-view>
+```
+
+The same happens with any event listeners directly attached to `RouterView`, they appear on `attrs` **so they can be easily passed down with one singe `v-bind`**.
+
+They also contain any _prop_ passed by using the `props` option in a route record:
+
+```js
+createRouter({
+  routes: [{ path: '/users/:id', component: User, props: true }],
```

Comment:
With `attrs` out of the way, I think we should re-consider what is passed there. I was thinking maybe it should pass the route params:

```vue
<router-view v-slot="{ Component, params }">
  <component :is="Component" v-bind="params"></component>
</router-view>
```

But then why not `query`? Why not other information of the current route?

Then I realized you already have access to the entire route object anyway:

```vue
<router-view v-slot="{ Component }">
  <component :is="Component"
    v-bind="$route.params"
    :dark-mode="$route.query.dark">
  </component>
</router-view>
```

But maybe we can still pass in the route object to make it rely less on magical global `$` properties:

```vue
<router-view v-slot="{ Component, route }">
  <component :is="Component"
    v-bind="route.params"
    :dark-mode="route.query.dark">
  </component>
</router-view>
```

This also makes it nicer to use in render functions:

``` js
import { RouterView } from 'vue-router'

return h(RouterView, ({ Component, route }) => h(Component, {
  ...route.params,
  darkMode: route.query.dark
}))
```

Reply:
added `route` to slot props